### ISSUE TITLE
Feature Request

### ISSUE THREAD

**Issue by geogolem** [batocera-linux/batocera-emulationstation#1701]
### Issue description

I'm not sure if this is the right place to open this issue.. but maybe you can provide some feedback and/or point me in the right direction..

First of all I am using Retrobat - but some of the discussion could also be applied to Batocera as well.

In Retrobat there is an option to autoconfigure the controllers which seems to be pretty useful BUT, its not easy to setup all my controllers this way in retrobat/emulation station so I prefer to turn that off and use retroarchs autoconfig. I can get all my controllers working no problem in retroarch using the most recent SDL2... 

The problem is it is not straightforward to map a 6 button Genesis controller or an n64 NSO controller in retrobat in such a way that it will then work properly in retroarch via retrobats autoconfig option.. so I disabled the autoconfig option and have all my controllers worker great inside retroarch.. I suppose if I could properly map everything in Retrobat I could re-enable the autoconfig and it would work fine too..but right now I cant.. (maybe the latest version of SDL2.dll would also resolve some of the issues I was having, but I am not sure)...

Anyway, even it might also be possible to map everything by manually editing the es_input.cfg file as well.. but for now I have disabled autoconfig in retrobat.

Having said all that.. my feature request is more related to the UI buttons.. I know there are options to show A/B or based on position.. I use the "based on position" buttons but the feature request is regarding the ability of retrobat to detect which controller is being used.. i.e. when navigating retrobat why don't the UI elements for the buttons change depending on the active controller... i.e. retrobat definitely can tell the difference between all my controllers... so it would be awesome if it could change the image being shown in realtime depending on the active controller.

**Comment by geogolem** [comment]
This isn't complete but it is pretty awesome: https://opengameart.org/content/free-keyboard-and-controllers-prompts-pack
https://www.reddit.com/r/gamedev/comments/1z0zid/free_keyboard_and_controllers_prompts_pack/

**Comment by geogolem** [comment]
maybe it could also be some kind of function.. i.e. tap hotkey on a controller, boom, the GUI changes to show the help icons corresponding to that controller..  pair another controller.. tap hotkey on it.. boom the GUI changes to show the help icons corresponding to that controller.. 

Of course, in this way, the user already needs to know what the hotkey (or other button combination/trigger) is for that specific controller.. I think it is better, to just always show the help icons corresponding to Port1/Player1 controller type... or whatever the last controller that had a button press.. i.e. then you can just tap left/right etc. to refresh the GUI with the helper icons corresponding to the controller... I know its sort of nitpicky.. but because I use different controllers, I always get confused which button is back vs. select and i often accidentally hit the wrong button.. Then the help icons actually can make things worse, since they dont always correspond to the controller i am using etc. lol

**Comment by geogolem** [comment]
so i started sort of wowkring on this... 

I realize the thing is messy with the externs etc.. but those are mostly just temporary.. also disabled the theme application for these buttons for now .. but eventually will need to figure out the best way to handle everything..

https://github.com/geogolem/batocera-emulationstation/tree/UIControllerDetection

this does work decently though, at least for my NSO SNES, NSO N64 and PS4 dualshock controllers..

**Comment by geogolem** [comment]
I think another important thing to do. is to augment the GUI Button Mapper in retrobat to be able to handle other buttons, such as A, B, C, X, Y, Z on 6 button genesis controller or c buttons etc. on n64 controllers.. this way they coould be properly mappy in retrobat and automatically work in retroarch with that mapping.. right not, i can map them well enough to use them in the retrobat UI.. and as long as I have the autoconfig option in retrobat disabled, i can easily configure them all using retroarchs autoconfig thing with the latest SDL2...

I really like the idea of being able to configure everything in retrobat.. but i just guess these more complicated controllers haven't been addressed yet.

**Comment by geogolem** [comment]
so i thought about this more.. and I am going to try modifying the implementation, so that basically everything is specified and setup in the theme; however, instead of the themeSetting choosing one to use all the time.. the controllerName/GUID determines which entry the theme uses. Then the only change to Batocera, would be to provbide the controllerName/GUID to the applyTheme function.. and update the Helpers on keypress... everything else regarding images and specification can be handled by the theme itself..

**Comment by geogolem** [comment]
So I started sort of thinking about augmenting the Controller Mapping Component to make use of some of these things based on GUID..

And also (separately) to potentially resolve issues with button mapping 6 button genesis controller or n64 style controller (c buttons etc.) so that the autoconfig feature of retrobat can work and make those controllers work in retroarch --> right now, there is no way to properly map the those types of controllers in retrobat at all without manually modifying the es_input.cfg file.

Now here is the dilemma --> with SNES/Playstation style controllers using only the button mappings currently handled, it just happens to work out, that if you map the controller properly in retrobat, the autoconfiguration function makes the controller work as expected in retroarch.. With the N64 controller specifically, I can modify the es_input.cfg file so that the autoconfig will make the controller work properly in retroarch.. (you can see how this guy does it here: https://www.youtube.com/watch?v=ER4DXS64CkM)

but the issue is, that the mapping needed to make the controller work in retroarch, doesn't necessarily intuitively match well with using the controller in retrobat es... i.e. the buttons might change the volume..  or I prefer c left to behave like (snes y --> search) in es.. but if I map it that way it breaks the autoconfiguration in retroarch.

Ultimately my current solution is to totally disable retrobat's autoconfiguration and let retroarch use the files defined here: https://github.com/libretro/retroarch-joypad-autoconfig 
thus allowing me to map the controller in retrobat (es_input.cfg) how I would like without breaking retroarchs mappings..

This seems to be the only solution unless there were 2 possible mapping settings in retrobat.. The existing one which maps button presses to buttons (which should hopefully and usually be automatically configured via SDL2 etc. but newer versions of SDL2 handle more controllers, retrobat still uses an older SDL2.dll release) and one (which doesn't fully exist) that specifies what buttons do what in retrobat... (this doesnt fully exist -- but partially exists with options to swap A/B or X/Y etc.. maybe those swapping options can be removed in favor of a more general mapping thing)...

The problem with the Swapping A/B or X/Y options is that they are globally applied.. and not necessarily on a per controller type basis.. i.e. what happens if sometimes I use a SNES controller.. and other times use an XBOX controller. with the SNES controller, I might want the A/B swapped.. Maybe it would be better, to basically add another column to the button mapping page... so that retrobat functions can be bound to different "buttons"....

In other words: Currently the mapping feature in retrobat allows binding physically controller buttons to "predefined retrobat buttons" then the autoconfig basically passes this configuration onto retroarch.. but there is no way to specify on a controller by controller level what each "predefined retrobat button" does other than those swap x/y, swap a/b options which apply globally and not on a per controller level.

(I am mostly just sharing my thoughts here for feedback.. my own future reference).. I realize this may not be a high priority feature.. but being new to retrobat/batocera the biggest challenge seems to be understanding the controller mapping etc. I do like the idea of retrobat having a UI to map the controller and then passing it onto retroarch etc.

**Comment by geogolem** [comment]
I'm thinking to simplify this and actually just make it switch between the existing buttons, for psx, xbox, nintendo etc. depending on the active controller. in other words, it may not actually be necessary to add any more images etc. but simply modify the theme to have a helpSystem  for Auto.. and then have the actual helpSystem used be decided based on the vendor ID of the device..

**Comment by geogolem** [comment]
https://reddit.com/r/batocera/comments/1cojgvo/change_buttons/
